retworkx-core/src/centrality.rs

```diff
@@ -297,3 +294,70 @@ where
         sigma,
     }
 }
+
+/// Compute the closeness centrality of all nodes in a graph.
+///
+/// Arguments:
+///
+/// * `graph` - The graph object to run the algorithm on
+/// * `wf_improved` - If True, scale by the fraction of nodes reachable.
+///
+/// # Example
+/// ```rust
+/// use retworkx_core::petgraph;
+/// use retworkx_core::centrality::closeness_centrality;
+///
+/// // Calculate the betweeness centrality of Graph
+/// let g = petgraph::graph::UnGraph::<i32, ()>::from_edges(&[
+///     (0, 4), (1, 2), (2, 3), (3, 4), (1, 4)
+/// ]);
+/// let output = closeness_centrality(&g, true);
+/// assert_eq!(
+///     vec![Some(1./2.), Some(2./3.), Some(4./7.), Some(2./3.), Some(4./5.)],
+///     output
+/// );
+///
+/// // Calculate the betweeness centrality of DiGraph
+/// let dg = petgraph::graph::DiGraph::<i32, ()>::from_edges(&[
+///     (0, 4), (1, 2), (2, 3), (3, 4), (1, 4)
+/// ]);
+/// let output = closeness_centrality(&dg, true);
+/// assert_eq!(
+///     vec![Some(0.), Some(0.), Some(1./4.), Some(1./3.), Some(4./5.)],
+///     output
+/// );
+/// ```
+pub fn closeness_centrality<G>(graph: G, wf_improved: bool) -> Vec<Option<f64>>
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + GraphBase<NodeId = NodeIndex>
+        + IntoEdges
+        + Visitable
+        + NodeCount
+        + IntoEdgesDirected,
+{
+    let max_index = graph.node_bound();
+    let mut betweenness: Vec<Option<f64>> = vec![None; max_index];
+    for node_s in graph.node_identifiers() {
+        let is = graph.to_index(node_s);
+        let map = dijkstra(Reversed(&graph), node_s, None, |_| 1);
+        let mut reachable_nodes_count = 0;
+        let mut dists_sum = 0;
+        for (_, &value) in map.iter() {
+            reachable_nodes_count += 1;
+            dists_sum += value;
+        }
+        if reachable_nodes_count == 1 {
+            betweenness[is] = Some(0.0);
+            continue;
+        }
+        betweenness[is] = Some((reachable_nodes_count - 1) as f64 / dists_sum as f64);
+        if wf_improved {
+            let node_count = graph.node_count();
+            betweenness[is] = betweenness[is]
+                .map(|c| c * (reachable_nodes_count - 1) as f64 / (node_count - 1) as f64);
+        }
+    }
```

Comment:
I think we can probably do this in parallel like we do for betweeness centrality if we just collect the node ids into a `Vec` and then use a parallel iterator over that vector. But we can always do this in a follow up PR too. Just having a serial implementation is more than enough to start.

Reply:
Your suggestions are very much appreciated. However, since I am still a beginner, I would like to deal only with serial implementations in this PR.